pkg/ebpf/ingress_node_firewall_loader.go

```diff
@@ -55,43 +59,295 @@ func NewIngNodeFwController() (*IngNodeFwController, error) {
+		return fmt.Errorf("cannot get map info: %v", err)
+	}
+	klog.Infof("Ingress node firewall map Info: %+v with FD %s", info, infc.objs.BpfMaps.IngressNodeFirewallTableMap.String())
+
+	// Convert IngressNodeFirewallRules into data that can be written to the BPF map.
+	// Build a map of valid ebpfKeys pointing to the ebpfRules that should be associated to them.
+	ebpfKeyToRules := make(map[BpfLpmIpKeySt]BpfRulesValSt)
+	for interfaceName, ingressRules := range ifaceIngressRules {
+		// Look up the network interface by name.
+		iface, err := net.InterfaceByName(interfaceName)
+		if err != nil {
+			return fmt.Errorf("lookup error for network iface %q: %s", interfaceName, err)
+		}
+		ifID := uint32(iface.Index)
+
+		// Convert each provided ingressRule into a mapping of potentially multiple keys (one for each CIDR)
+		// pointing to a flattened rule that can be written to the BPF map.
+		for _, rule := range ingressRules {
+			if ebpfKeys, ebpfRules, err := infc.makeIngressFwRulesMap(rule, ifID); err == nil {
+				for _, ebpfKey := range ebpfKeys {
+					ebpfKeyToRules[ebpfKey] = ebpfRules
+				}
+			} else {
+				return fmt.Errorf("failed to create map firewall rules: %v on if %d", err, ifID)
+			}
+		}
+	}
+
+	// Build a slice of desired keys - it's easier to iterate over this slice later.
+	var desiredKeys []BpfLpmIpKeySt
+	for desiredKey := range ebpfKeyToRules {
+		desiredKeys = append(desiredKeys, desiredKey)
+	}
+
+	// Get stale keys (= keys inside the eBPF map but not inside the currently desired ruleset). Those keys
+	// will be dropped in the next step.
+	staleKeys, err := infc.getStaleKeys(desiredKeys)
 	if err != nil {
-		return fmt.Errorf("Cannot get map info: %v on if %d", err, ifId)
+		return err
+	}
+
+	// Purge all stale keys from the eBPF map.
+	if err := infc.purgeKeys(staleKeys); err != nil {
+		klog.Infof("Purge keys operation encountered issues, err: %q", err)
 	}
-	klog.Infof("Ingress node firewall map Info: %+v with FD %s", info, objs.BpfMaps.IngressNodeFirewallTableMap.String())
 
-	if err := infc.makeIngressFwRulesMap(ingFireWallConfig, isDelete, ifId); err != nil {
-		return fmt.Errorf("Failed to create map firewall rules: %v on if %d", err, ifId)
+	// Add/update all keys. This is an idempotent action and non-existing keys are added whereas existing keys
+	// are updated.
+	if err := infc.addOrUpdateRules(ebpfKeyToRules); err != nil {
+		return err
 	}
 
+	// Generate ingress node few events.
 	infc.ingressNodeFwEvents()
 
 	return nil
 }
 
+// addOrUpdateRules is a small method containing this limited set of functionality to facilitate unit testing of
+// this part of code.
+func (infc *IngNodeFwController) addOrUpdateRules(ebpfKeyToRules map[BpfLpmIpKeySt]BpfRulesValSt) error {
+	for ebpfKey, ebpfRules := range ebpfKeyToRules {
+		log.Printf("Adding or updating ingress firewall rules for key %v", ebpfKey)
+		if err := infc.objs.BpfMaps.IngressNodeFirewallTableMap.Update(ebpfKey, ebpfRules, ebpf.UpdateAny); err != nil {
```

Comment:
this is the part u said it doesn't work on slack ? pls add FIXME here for me to investigate so I won't forget about it

Reply:
done